vwifi.c

```diff
@@ -39,62 +37,77 @@ struct owl_context {
 };
 
 /* SME stands for "station management entity" */
 enum sme_state { SME_DISCONNECTED, SME_CONNECTING, SME_CONNECTED };
 
 /* Virtual interface which is pointed by netdev_priv(). Fields in the
  * structure are interface-dependent.
+ * Every interface owns a owl_vif, no matter the interface mode
+ * (STA, AP, Ad-hoc...).
  */
 struct owl_vif {
     struct wireless_dev wdev;
     struct net_device *ndev;
     struct net_device_stats stats;
 
+    size_t ssid_len;
     /* Currently connected BSS id */
     u8 bssid[ETH_ALEN];
     u8 ssid[IEEE80211_MAX_SSID_LEN];
-    /* For the case the STA is going to roam to another BSS */
-    u8 req_bssid[ETH_ALEN];
-    u8 req_ssid[IEEE80211_MAX_SSID_LEN];
-    struct cfg80211_scan_request *scan_request;
-    enum sme_state sme_state;  /* connection information */
-    unsigned long conn_time;   /* last connection time to a AP (in jiffies) */
-    unsigned long active_time; /* last tx/rx time (in jiffies) */
-    u16 disconnect_reason_code;
-
-    struct mutex lock;
-    struct timer_list scan_timeout;
-    struct work_struct ws_connect, ws_disconnect;
-    struct work_struct ws_scan, ws_scan_timeout;
-    struct list_head rx_queue; /* Head of received packet queue */
 
-    /* List entry for maintaining multiple private data of net_device in
-     * owl_context.vif_list.
+    /* Head of received packet queue */
+    struct list_head rx_queue;
+    /* Store all owl_vif which is in the same BSS (AP will be the head). */
+    struct list_head bss_list;
+    /* List entry for maintaining all owl_vif, which can be access via
+     * owl->vif_list.
      */
     struct list_head list;
-};
-
-/* AP information table entry */
-struct ap_info_entry_t {
-    struct hlist_node node;
-    u8 bssid[ETH_ALEN];
-    char ssid[IEEE80211_MAX_SSID_LEN];
-};
 
-static char *ssid_list = DEFAULT_SSID_LIST;
-module_param(ssid_list, charp, 0644);
-MODULE_PARM_DESC(ssid_list, "Self-defined SSIDs.");
+    struct mutex lock;
 
-static int interfaces = 2;
-module_param(interfaces, int, 0444);
-MODULE_PARM_DESC(interfaces, "Number of virtual interfaces.");
+    /* Split logic for STA and AP mode */
+    union {
+        /* Structure for STA mode */
+        struct {
+            /* For the case the STA is going to roam to another BSS */
+            u8 req_ssid[IEEE80211_MAX_SSID_LEN];
+
+            struct cfg80211_scan_request *scan_request;
+            enum sme_state sme_state; /* connection information */
+            unsigned long
+                conn_time; /* last connection time to a AP (in jiffies) */
+            unsigned long active_time; /* last tx/rx time (in jiffies) */
+            u16 disconnect_reason_code;
+
+            struct timer_list scan_timeout;
+            struct work_struct ws_connect, ws_disconnect;
+            struct work_struct ws_scan, ws_scan_timeout;
+
+            /* For quickly finding the AP */
+            struct owl_vif *ap;
+        };
+        /* Structure for AP mode */
+        struct {
```

Comment:
Why another struct which contains `struct lisy_head` only?

Reply:
Although currently the struct only contains a field, but in the future it may accommodate more members related to AP.